src/pykeen/evaluation/rank_based_evaluator.py

```diff
@@ -597,6 +600,136 @@ def finalize(self) -> RankBasedMetricResults:  # noqa: D102
+        tail_negatives: Optional[torch.LongTensor] = None,
+        **kwargs,
+    ):
+        """
+        Initialize the evaluator.
+
+        :param evaluation_factory:
+            the factory with evaluation triples
+        :param head_negatives: shape: (num_triples, num_negatives)
+            the entity IDs of negative samples for head prediction for each evaluation triple
+        :param tail_negatives: shape: (num_triples, num_negatives)
+            the entity IDs of negative samples for tail prediction for each evaluation triple
+        :param kwargs:
+            additional keyword-based arguments passed to RankBasedEvaluator.__init__
+        """
+        super().__init__(**kwargs)
+        if head_negatives is None and tail_negatives is None:
+            # default for inductive LP by [teru2020]
+            num_negatives = num_negatives or 50
+            logger.info("Sampling negatives")
+            if num_negatives > evaluation_factory.num_entities:
+                raise ValueError("Cannot use more negative samples than there are entities.")
+            head_negatives, tail_negatives = sample_negatives(
+                evaluation_triples=evaluation_factory.mapped_triples,
+                additional_filter_triples=additional_filter_triples,
+                num_entities=evaluation_factory.num_entities,
+                num_samples=num_negatives,
+            )
+        elif head_negatives is None or tail_negatives is None:
+            raise ValueError("Either both, head and tail negatives must be provided, or none.")
+
+        # verify input
+        for negatives in (head_negatives, tail_negatives):
+            if negatives.shape[0] != evaluation_factory.num_triples:
+                raise ValueError(f"Negatives are in wrong shape: {negatives.shape}")
+        self.triple_to_index = {(h, r, t): i for i, (h, r, t) in enumerate(evaluation_factory.mapped_triples.tolist())}
+        self.negative_samples = {
+            SIDE_HEAD: head_negatives,
+            SIDE_TAIL: tail_negatives,
+        }
+        self.num_entities = evaluation_factory.num_entities
+
+    def _update_ranks_(
+        self,
+        true_scores: torch.FloatTensor,
+        all_scores: torch.FloatTensor,
+        side: str,
+        hrt_batch: MappedTriples,
+    ) -> None:  # noqa: D102
+        # TODO: do not require to compute all scores beforehand
+        triple_indices = [self.triple_to_index[h, r, t] for h, r, t in hrt_batch.cpu().tolist()]
+        negative_entity_ids = self.negative_samples[side][triple_indices]
+        negative_scores = all_scores[
+            torch.arange(hrt_batch.shape[0], device=hrt_batch.device).unsqueeze(dim=-1),
+            negative_entity_ids,
+        ]
+        # super.evaluation assumes that the true scores are part of all_scores
+        scores = torch.cat([true_scores, negative_scores], dim=-1)
+        super()._update_ranks_(true_scores=true_scores, all_scores=scores, side=side, hrt_batch=hrt_batch)
```

Comment:
I am not sure about this line - it will change the total number of entities from `total_entities` to `num_negatives + 1` because the super() method only gets the num_entities from the all_scores shape

Reply:
this case now catches this error 
4d2e6e4

Reply:
it is fixed in 1a5bddb